src/persist/src/golden_test.rs

```diff
@@ -0,0 +1,227 @@
+    // more interesting events.
+    config.storage_unavailable = 0;
+    config.storage_available = 0;
+    config.start_weight = 0;
+    config.stop_weight = 0;
+    let g = Generator::new(seed, config);
+    let reqs = g.into_iter().take(100).collect::<Vec<_>>();
+
+    let golden = golden_state(DATAZ)?;
+    let (current, raw_blobs) = current_state(&reqs)?;
+
+    if golden != current {
+        for req in reqs {
+            log::debug!("req {:?}", req);
+        }
+        log::info!("current impl blob state: {}", raw_blobs);
+        assert_eq!(golden, current);
+    }
+
+    Ok(())
+}
+
+fn golden_state(blob_json: &str) -> Result<PersistState, Error> {
+    let mut blob = MemBlob::new_no_reentrance("");
+    if let Err(err) = Blobs::deserialize_to(blob_json, &mut blob) {
+        log::error!("error deserializing golden: {}", err);
+    }
+    let mut persist = runtime::start(ErrorLog, blob, &MetricsRegistry::new())?;
+    let state = PersistState::slurp_from(&persist)?;
+    persist.stop()?;
+    Ok(state)
+}
+
+fn current_state(reqs: &[Input]) -> Result<(PersistState, String), Error> {
+    let reg = Arc::new(Mutex::new(MemRegistry::new()));
+    let runtime_reg = reg.clone();
+    let mut persist = Direct::new(move |unreliable| {
+        let blob = runtime_reg
+            .lock()?
+            .blob("", LockInfo::new_no_reentrance("".into()))?;
+        let blob = UnreliableBlob::from_handle(blob, unreliable);
+        runtime::start(ErrorLog, blob, &MetricsRegistry::new())
+    })?;
+    for req in reqs.iter() {
+        let _ = persist.run(req.clone());
+    }
+    let persist_state = PersistState::slurp_from(&persist.persister)?;
+    persist.finish();
+
+    let mut blob = reg
+        .lock()?
+        .blob("", LockInfo::new_no_reentrance("".into()))?;
+    let raw_blobs = Blobs::serialize_from(&blob)?;
+    blob.close()?;
+    Ok((persist_state, raw_blobs))
+}
+
+#[derive(Serialize, Deserialize)]
+struct Blobs {
+    blobs: Vec<(String, Vec<u8>)>,
+}
+
+impl Blobs {
+    fn deserialize_to<B: Blob>(blob_json: &str, blob: &mut B) -> Result<(), Error> {
+        let blobs: Blobs =
+            serde_json::from_str(blob_json).map_err(|err| Error::from(err.to_string()))?;
+        for (key, val) in blobs.blobs.iter() {
+            blob.set(&key, val.to_owned(), false)?;
+        }
+        Ok(())
+    }
+
+    fn serialize_from(blob: &MemBlob) -> Result<String, Error> {
+        let blobs = Blobs {
+            blobs: blob.all_blobs()?,
+        };
+        let blob_json =
+            serde_json::to_string(&blobs).map_err(|err| Error::from(err.to_string()))?;
+        Ok(blob_json)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct PersistState {
+    streams: Vec<PersistStreamState>,
+}
+
+impl PersistState {
+    fn slurp_from(persist: &RuntimeClient) -> Result<Self, Error> {
+        // TODO: There is currently no way to get the set of active streams out
```

Comment:
I imagine this is a TODO for the future, not for this PR?

Reply:
Exactly, yes. I personally use TODO for things to fix later and WIP for things to fix before merging.